src/metabase/cache_database.clj

```diff
@@ -0,0 +1,75 @@
+(ns metabase.cache-database
+  "The logic for doing DB and Table syncing itself.
+   TODO: move this file under the sync_database directory"
+  (:require [clojure.tools.logging :as log]
+            [metabase
+             [driver :as driver]
+             [events :as events]
+             [util :as u]]
+            [metabase.models.table :as table]
+            [metabase.query-processor.interface :as i]
+            [metabase.sync-database.cached-values :as cached-values]
+            [toucan.db :as db]))
+
+(defonce ^:private currently-syncing-db-ids
+  (atom #{}))
+
+(defn- cache-database-with-tracking! [driver database]
+  (let [start-time (System/nanoTime)
+        tracking-hash (str (java.util.UUID/randomUUID))]
+    (log/info (u/format-color 'magenta "Syncing %s database '%s'..." (name driver) (:name database)))
+    (events/publish-event! :database-analysis-begin {:database_id (:id database) :custom_id tracking-hash})
```

Comment:
so for caching a "database data shape" we publish an event called `database-analysis-begin`? Does that make any sense?